### PR TITLE
[dualtor][DualTorIO] Fix thread hang waiting for io_ready

### PR DESCRIPTION
**Pull request by lolyu**
### Description of PR

For the main thread, it will wait for the packet sender to set event
`io_ready`. If any error occurs within the sender thread and it fails to
set the event, the main thread will hang waiting for the event.

1. Add a error handler to InterruptableThread so it could call this
handelr as callback in case of any errors.

2. Make the sender thread an instance of InterruptableThread and add a
error handler to set `io_ready` to let main thread continue to run.

Signed-off-by: Longxiang Lyu <lolv@microsoft.com>

Summary:
Fixes # (issue)

### Type of change

- [x] Bug fix
- [ ] Testbed and Framework(new/improvement)
- [ ] Test case(new/improvement)

#### What is the motivation for this PR?
For the main thread, it will wait for the packet sender to set event
`io_ready`. If any error occurs within the sender thread and it fails to
set the event, the main thread will hang waiting for the event.

#### How did you do it?
1. Add a error handler to `InterruptableThread` so it could call this
handelr as callback in case of any errors.

2. Make the sender thread an instance of `InterruptableThread` and add a
error handler to set `io_ready` to let main thread continue to run.

#### How did you verify/test it?
* demo test
```python
import logging
import functools
import pdb
import random
import time

from tests.common.dualtor.dual_tor_utils import upper_tor_host
from tests.common.dualtor.dual_tor_utils import lower_tor_host
from tests.common.dualtor.data_plane_utils import send_t1_to_server_with_action
from tests.common.dualtor.mux_simulator_control import set_drop
from tests.common.dualtor.mux_simulator_control import set_output
from tests.common.dualtor.mux_simulator_control import toggle_all_simulator_ports_to_upper_tor
from tests.common.fixtures.ptfhost_utils import run_icmp_responder
from tests.common.fixtures.ptfhost_utils import run_garp_service
from tests.common.fixtures.ptfhost_utils import change_mac_addresses             # lgtm[py/unused-import]
from tests.common.fixtures.ptfhost_utils import copy_ptftests_directory


def test_active_link_drop_downstream_active(
    upper_tor_host,
    lower_tor_host,
    send_t1_to_server_with_action,
    set_drop,
    set_output,
    toggle_all_simulator_ports_to_upper_tor,
    tor_mux_intfs
):
    def timer(func):
        def decorated(*args, **kargs):
            start_time = time.time()
            logging.debug("Start time: %s", start_time)
            ret = func(*args, **kargs)
            end_time = time.time()
            logging.debug("End time: %s", end_time)
            return ret
        return decorated

    pdb.set_trace()
    intf = random.choice(tor_mux_intfs)
    logging.debug(intf)
    io_obj = send_t1_to_server_with_action(
        upper_tor_host,
        tor_vlan_port=intf,
        action=functools.partial(timer(set_drop), intf, ["upper_tor"])
    )
    set_output(intf, ["upper_tor", ])
```
```
dualtor/test_link_drop.py::test_active_link_drop_downstream_active PASSED                                                                                                                      [100%]

===================================================================================== 1 passed in 609.30 seconds =====================================================================================
```

### CODE REVIEW

**Review by wangxin — APPROVED** [review]

File: tests/common/dualtor/dual_tor_io.py [hunk 1]
```diff
@@ -180,9 +181,10 @@ def generate_from_t1_to_server(self):
         if self.tor_vlan_intf:
             # If destination VLAN intf is specified,
             # use only the connected server
-            server_ip_list = [self.ptf_intf_to_server_ip_map[
-                self.tor_vlan_intf
-                ]]
+            ptf_port = self.tor_to_ptf_intf_map[self.tor_vlan_intf]
+            server_ip_list = [
+                self.ptf_intf_to_server_ip_map[ptf_port]
```
Comment (vaibhavhd) [inline tests/common/dualtor/dual_tor_io.py:186]: Why was this changed to `ptf_port`? Can it not be kept to the instance variable `self.tor_vlan_intf`?

File: tests/common/utilities.py [hunk 2]
```diff
@@ -101,6 +105,8 @@ def run(self):
             threading.Thread.run(self)
         except Exception:
             self._e = sys.exc_info()
+            if self.error_handler:
```
Comment (vaibhavhd) [inline tests/common/utilities.py:108]: If a call to `set_error_handler` is not made, this Exception might throw a Python error where `error_handler` attribute is not present in `InterruptableThread`.

File: tests/common/dualtor/dual_tor_io.py (hunk 1 above)
Comment (theasianpianist) [reply tests/common/dualtor/dual_tor_io.py:186]: Looks like there was an error in my original code, `self.tor_vlan_intf` should never exist in `self.ptf_intf_to_server_ip_map`. Good catch @lolyu

File: tests/common/dualtor/dual_tor_io.py [hunk 3]
```diff
@@ -327,8 +330,8 @@ def send_and_sniff(self, sender, sniffer):
         """
         @summary: This method starts and joins two background threads in parallel: sender and sniffer
         """
-        self.sender_thr = threading.Thread(target=sender)
-        self.sniff_thr = threading.Thread(target=sniffer)
+        self.sender_thr = InterruptableThread(target=sender)
```
Comment (theasianpianist) [inline tests/common/dualtor/dual_tor_io.py:333]: Do we need to call `set_error_handler` for these two?

File: tests/common/dualtor/dual_tor_io.py (hunk 1 above)
Comment (vaibhavhd) [reply tests/common/dualtor/dual_tor_io.py:186]: Oh! So I guess the previous code did not hit an error because `self.ptf_intf_to_server_ip_map[self.tor_vlan_intf]]` still returned some IPs based on ptf_port that was same as tor_vlan_intf.

However, this should have failed/caught in link down testcases, where a specific `tor_vlan_intf` is specified. Are we missing any other issue?

Comment (theasianpianist) [reply tests/common/dualtor/dual_tor_io.py:186]: The link down test cases take down all interfaces, not just a single one. So I don't believe this portion of code was ever run, since all the test cases so far applied any disruptive actions to all ports. There was a similar error in `generate_server_to_t1` but it looks like @lolyu has already fixed it below

File: tests/common/dualtor/dual_tor_io.py [hunk 4]
```diff
@@ -256,22 +258,23 @@ def generate_from_server_to_t1(self):
...
-                                               ]
-                           )
-                   )
+        if self.tor_vlan_intf is None:
+            src_mac = 'random'
+            src_ip = 'random'
+        else:
+            ptf_port = self.tor_to_ptf_intf_map[self.tor_vlan_intf]
+            src_mac = ptf_intf_to_mac_map[ptf_port]
+            src_ip = self.tor_to_ptf_intf_map[ptf_port]
```
Comment (theasianpianist) [inline tests/common/dualtor/dual_tor_io.py:267]: I believe there is an additional dictionary lookup to get the server IP, in `self.ptf_intf_to_server_ip_map`

File: tests/common/dualtor/dual_tor_io.py (hunk 3 above)
Comment (lolyu) [reply tests/common/dualtor/dual_tor_io.py:333]: Yes, looks like we should put a `set_error_handler` for the sniffer thread so set the `sniffer_started` event to free the sender thread, will update.

File: tests/common/utilities.py (hunk 2 above)
Comment (bingwang-ms) [reply tests/common/utilities.py:108]: Good catch.

**Comment by lolyu** [comment]
Updated
```
dualtor/test_link_drop.py::test_active_link_drop_downstream_active PASSED                                                                                                                      [100%]

===================================================================================== 1 passed in 515.01 seconds =====================================================================================
```

File: tests/common/dualtor/dual_tor_io.py (hunk 3 above)
Comment (theasianpianist) [reply tests/common/dualtor/dual_tor_io.py:333]: Does the sender thread also need an error handler? `run_test` in `data_plane_utils.py` waits for the sender to set `io_ready_event` so any errors here might cause a the main thread to hang.

**Comment by lolyu** [comment]
Could you please take a review again @theasianpianist, @vaibhavhd?

File: tests/common/dualtor/dual_tor_io.py (hunk 3 above)
Comment (lolyu) [reply tests/common/dualtor/dual_tor_io.py:333]: Updated, thanks!

**Comment by lolyu** [comment]
```
dualtor/test_link_drop.py::test_active_link_drop_downstream_active PASSED                                                                                                                                                                                              [100%]

========================================================================================================================= 1 passed in 555.27 seconds =========================================================================================================================
```

**Review by vaibhavhd — APPROVED** [review]

**Review by theasianpianist — APPROVED** [review]
lgtm, thanks for the fixes!
